curriculum/challenges/english/07-scientific-computing-with-python/learn-special-methods-by-building-a-vector-space/65f9b9710cca621244d3bde1.md

```diff
@@ -0,0 +1,59 @@
+---
+id: 65f9b9710cca621244d3bde1
+title: Step 35
+challengeType: 20
+dashedName: step-35
+---
+
+# --description--
+
+Replace `pass` with a `raise` statement that raises a `TypeError` and use the following f-string to provide a custom error message: `f"Addition is not defined between '{self.__class__.__name__}' and '{other.__class__.__name__}'"`.
```

Comment:
Was returning `NotImplemented` instead of raising exception considered? 

Exception limits ability to use this class with other classes, by assuming they will not know how to perform certain operation.

On the other hand returning `NotImplemented` is like saying: _I don't know how to do this, ask the other party._ If neither side knows what to so, it will still end up with `TypeError`.

Reply:
using `NotImplemented` resulted in being able to add together a R2Vector and a R3Vector instances

Reply:
The problem was `isinstance` and I didn't realize it... If we use `self.__class__ != other.__class__` instead of `not isinstance(other, self.__class__)`, using `NotImplemented` is perfectly fine.

Reply:
This is technically correct with the current classes structure, because `R3Vector` is subclass of `R2Vecor`. `isinstance` checks also parent classes. It might look like that's rather strange, but it's good thing - it means classes do not need to have all methods rewritten by default, to make them work between child and parent.

This could be corrected with a more appropriate structure - new `Vector` class would be parent of both `R2Vector` and `R3Vector`. Effectively `R2Vector` and `R3Vector` would not be _directly_ related. So `isinstance(R3Vector(), R2Vector)` is `False`.

Reply:
Personally, I like the idea of having a `Vector` class and two child classes `R2Vector` and `R3Vector`. It's more appropriate, as you said, and we can use `isinstance` without any issues.

This means I need to move every method which is currently in `R2Vector` inside `Vector`, except `__init__`.
The result would be:
```py
class Vector:
   def norm(self):
        return sum(val**2 for val in vars(self).values())**0.5
   ...

   def __ge__(self, other):
        return not self < other

class R2Vector(Vector):
    def __init__(self, *, x, y):
        self.x = x
        self.y = y

class R3Vector(Vector):
    def __init__(self, *, x, y, z):
        self.x = x
        self.y = y
        self.z = z
        
    def cross(self, other):
        if not isinstance(other, self.__class__):
            return NotImplemented
        kwargs = {
            'x': self.y * other.z - self.z * other.y,
            'y': self.z * other.x - self.x * other.z,
            'z': self.x * other.y - self.y * other.x
        }
        
        return self.__class__(**kwargs)
```
For that I need to remove the `super` call, which by the way was just to show its existence.
What do you think?

Reply:
Although I like the idea of introducing a single parent class, considering the time needed to implement all the changes, I think for now the project will be fine as it is.

We can always modify the structure later, if we decide it's better.

Reply:
Some way to keep `super` relevant could be found... 😛 

```python
class Vector:
    def __init__(self, **kwargs):
        for keyword, value in kwargs.items():
            setattr(self, keyword, value)

class R2Vector(Vector):
    def __init__(self, *, x, y):
        super().__init__(x=x, y=y)


class R3Vector(Vector):
    def __init__(self, *, x, y, z):
        super().__init__(x=x, y=y, z=z)
```

Reply:
This is clever and elegant. We might want to implement it in a future version of this project :clap: